src/app/blog/0rtt-api/page.mdx

```diff
@@ -0,0 +1,397 @@
+Node ID: PublicKey(1c64a26558be5f114375977a32d411ba5610577d4935ecb3f00b015320c32e78)
+Ticket: nodeaa...
+```
+
+## Connect side
+
+On the connect side, we just need to specify the node ticket to connect to and optionally the number of rounds.
+
+We set a debug level to see what's going on:
+
+```
+RUST_LOG=0rtt=trace cargo run --release --example 0rtt nodeaa...
+2025-05-22T15:10:02.005378Z TRACE 0rtt: 0-RTT not possible from our side
+2025-05-22T15:10:02.270286Z DEBUG 0rtt: round 0: 266358 us
+2025-05-22T15:10:02.272468Z TRACE 0rtt: 0-RTT possible from our side
+2025-05-22T15:10:02.402081Z DEBUG 0rtt: round 1: 131681 us
+2025-05-22T15:10:02.404558Z TRACE 0rtt: 0-RTT possible from our side
+2025-05-22T15:10:02.533311Z DEBUG 0rtt: round 2: 130064 us
+2025-05-22T15:10:02.534596Z TRACE 0rtt: 0-RTT not possible from our side
+2025-05-22T15:10:02.792593Z DEBUG 0rtt: round 3: 259101 us
+2025-05-22T15:10:02.793975Z TRACE 0rtt: 0-RTT possible from our side
+2025-05-22T15:10:02.924554Z DEBUG 0rtt: round 4: 131883 us
+2025-05-22T15:10:02.926120Z TRACE 0rtt: 0-RTT possible from our side
+2025-05-22T15:10:03.055240Z DEBUG 0rtt: round 5: 130586 us
+2025-05-22T15:10:03.056524Z TRACE 0rtt: 0-RTT not possible from our side
+2025-05-22T15:10:03.316197Z DEBUG 0rtt: round 6: 260394 us
+2025-05-22T15:10:03.317830Z TRACE 0rtt: 0-RTT possible from our side
+
+```
+
+Now this is weird.
+
+The initial connection is not 0-RTT, which is expected. The two nodes haven't ever talked, so there is no pre shared secret they can use to communicate.
+
+The next two rounds are 0-RTT, as expected, and have roughly the expected roundtrip times.
+
+But then we get a non 0-RTT connection on the 4th round. What's going on?
+
+The issue is that 0-RTT is done using *single-use* tickets. *After* each completed handshake, you get a number of these tickets, by default 2. And as soon as the tickets are used up, the client can not even *attempt* to do a 0-rtt request because it has run out of tickets to use.
+
+In the code so far we close the connection immediately after receiving the user data response. So we don't wait for the full handshake to complete, and also don't get any additional tickets.
+
+```rust
+pingpong_0rtt(connecting, i).await?
+connection.close(0u8.into(), b"");
+```
+
+## Waiting for new tickets
+
+There are two ways to solve this. The tickets for 0-RTT connections are single use in quinn, as recommended in the TLS spec: https://datatracker.ietf.org/doc/html/rfc8446#appendix-C.4 . We can get new tickets via a connection that was initiated with 0-RTT, but only if we allow time for the handshake to complete.
+
+There is currently no quinn API to wait for the handshake to complete or to wait for session tickets. But we can just sleep for two times the current round trip time before closing the connection. This is a crude way to allow for enough time for the two `NewSessionTicket`s to be received.
```

Comment:
I think here as well this would be fine without mentioning Quinn.  If you want to make up excuses for us not adding an API you can invoke Rustls.